src/compat/floret/indexer.rs

```diff
@@ -0,0 +1,57 @@
+use std::io::Cursor;
+
+use murmur3::murmur3_x64_128;
+use smallvec::smallvec;
+
+use crate::subword::{Indexer, NGramVec, StrWithCharLen};
+
+/// floret subword indexer.
+#[derive(Clone, Copy, Debug, Eq, PartialEq)]
+pub struct FloretIndexer {
+    n_buckets: u64,
+    seed: u32,
+    n_hashes: u32,
+}
+
+impl FloretIndexer {
+    pub fn new(n_buckets: u64, n_hashes: u32, seed: u32) -> Self {
+        assert!(
+            n_hashes > 0 && n_hashes <= 4,
+            "Floret indexer needs 1 to 4 hashes, got {}",
+            n_hashes
+        );
+
+        Self {
+            n_buckets,
+            n_hashes,
+            seed,
+        }
+    }
+}
+
+impl Indexer for FloretIndexer {
+    fn index_ngram(&self, ngram: &StrWithCharLen) -> NGramVec {
+        let hash = murmur3_x64_128(&mut Cursor::new(ngram.as_bytes()), self.seed)
+            .expect("Murmur hash failed");
+
+        let hash_array: [u32; 4] = unsafe { std::mem::transmute(hash) };
```

Comment:
Feel free to ignore this if you think it's too pedantic, but I think a small comment regarding safety is nice to have next to an unsafe block.

On a different note, the mips64 tests failed while other platforms worked, does this `transmute` lead to endianness issues?

Reply:
We should probably get rid of the `unsafe` altogether. A simple bit shift approach get compiled away:

https://godbolt.org/z/KjKnKbvPM

But I have to investigate the ramifications a bit wrt. murmur3 and endianness.

Reply:
I compared some outputs of the Murmur3 implementation and they are the same for any inputs I tried on mips via cross and on native x86_64:

~~~Rust
use std::io::Cursor;

fn main() {
    let inputs = vec!["test", "another", "asd"];
    for input in inputs {
        println!(
            "{}: {}",
            input,
            murmur3::murmur3_x64_128(&mut Cursor::new(input.as_bytes()), 2166136261).unwrap()
        );
    }
    // mips
    // test: 143416175659979471159943107701931823025
    // another: 19454989077232334339732095908570983208
    // asd: 49449963734955912795853045277450408020

    // x86_64
    // test: 143416175659979471159943107701931823025
    // another: 19454989077232334339732095908570983208
    // asd: 49449963734955912795853045277450408020

    let input = (0..255).cycle().take(1024).collect::<Vec<u8>>();
    println!(
        "{}",
        murmur3::murmur3_x64_128(&mut Cursor::new(input), 2166136261).unwrap()
    );
    // x86_64: 121525936877850996829446231388818626050
    // mips: 121525936877850996829446231388818626050
}
~~~